### PR TITLE
allow better SIMD paralelization of SHA-1

### PR DESCRIPTION
**Pull request by wilx**
The following change uses equivalent algorithm modifications recommended in <https://software.intel.com/en-us/articles/improving-the-performance-of-the-secure-hash-algorithm-1/> (via [Wikipedia](http://en.wikipedia.org/wiki/SHA-1)) to allow better SIMD paralelization. With GCC 4.9.1 on AMD64 and `-O3`, I can see the second loop being vectorized with SSE instructions, which does not happen without the modification. I have not done any benchmarking. I do not suggest that you merge this pull request as it is. I just think this might be something for you to consider and maybe incorporate into Botan.

This is a duplicate of #50 for master branch.

### CODE REVIEW

**Comment by wilx** [comment]
The following change uses equivalent algorithm modifications recommended in <https://software.intel.com/en-us/articles/improving-the-performance-of-the-secure-hash-algorithm-1/> (via [Wikipedia](http://en.wikipedia.org/wiki/SHA-1)) to allow better SIMD paralelization. With GCC 4.9.1 on AMD64 and `-O3`, I can see the second loop being vectorized with SSE instructions, which does not happen without the modification. I have not done any benchmarking. I do not suggest that you merge this pull request as it is. I just think this might be something for you to consider and maybe incorporate into Botan.

**Comment by webmaster128** [comment]
Thanks @wilx for the contribution. Since the botan tests include 70+ sha1 tests, the code is likely to be correct. Could you provide a benchmark program (e.g. a main.cpp utilizing the [STL timer](http://en.cppreference.com/mwiki/index.php?title=Special%3ASearch&search=chrono&fulltext=1) and Botans sample data in `src/tests/data/hash/sha1.vec`)? Then we know whether it is worth further reviewing effort.

**Comment by wilx** [comment]
# TL;DR

I withdraw the change. I can consistently measure about 2.7 % slower execution with the change.

# Long version

So I have done some testing. I have used Botan's hash test and I have deleted all `*.vec` files but the SHA1. I have modified the `test_hash.cpp` like this:

~~~~
--- a/src/tests/test_hash.cpp
+++ b/src/tests/test_hash.cpp
@@ -11,6 +11,7 @@
 #include <botan/hex.h>
 #include <iostream>
 #include <fstream>
+#include <chrono>
 
 using namespace Botan;
 
@@ -54,11 +55,24 @@ size_t hash_test(const std::string& algo,
          }
 
       // Test to make sure clear() resets what we need it to
+
+      typedef std::chrono::high_resolution_clock hr_clock;
+      typedef std::chrono::duration<double, std::ratio<1>> sec_dur_type;
+
+      hr_clock::time_point start = hr_clock::now ();
+      for (int i = 0; i != 10000000; ++i)
+          {
+          hash->update(in);
+          }
+      hr_clock::time_point end = hr_clock::now ();
+      hr_clock::duration diff = end - start;
+      double diff_seconds = sec_dur_type (diff).count ();
+      std::cout << algo << '/' << in_hex << ": " << diff_seconds << "s\n";
+
       hash->update("some discarded input");
       hash->clear();
 
       hash->update(in);
-
       h = hash->final();
 
       if(h != hex_decode_locked(out_hex))
~~~~

The slowdown with the code as it is in this pull request was even bigger. In the end I have ended up unrolling the second loop as well like the original code does:

~~~~
--- a/src/lib/hash/sha1/sha160.cpp
+++ b/src/lib/hash/sha1/sha160.cpp
@@ -95,9 +95,16 @@ void SHA_160::compress_n(const byte input[], size_t blocks)
          W[j+7] = rotate_left((W[j+4] ^ W[j-1] ^ W[j- 7] ^ W[j- 9]), 1);
          }
 
-      for(size_t j = 32; j != 80; ++j)
+      for(size_t j = 32; j != 80; j += 8)
          {
-         W[j  ] = rotate_left((W[j-6] ^ W[j-16] ^ W[j-28] ^ W[j-32]), 2);
+         W[j  ] = rotate_left((W[j-6+0] ^ W[j-16+0] ^ W[j-28+0] ^ W[j-32+0]), 2);
+         W[j+1] = rotate_left((W[j-6+1] ^ W[j-16+1] ^ W[j-28+1] ^ W[j-32+1]), 2);
+         W[j+2] = rotate_left((W[j-6+2] ^ W[j-16+2] ^ W[j-28+2] ^ W[j-32+2]), 2);
+         W[j+3] = rotate_left((W[j-6+3] ^ W[j-16+3] ^ W[j-28+3] ^ W[j-32+3]), 2);
+         W[j+4] = rotate_left((W[j-6+4] ^ W[j-16+4] ^ W[j-28+4] ^ W[j-32+4]), 2);
+         W[j+5] = rotate_left((W[j-6+5] ^ W[j-16+5] ^ W[j-28+5] ^ W[j-32+5]), 2);
+         W[j+6] = rotate_left((W[j-6+6] ^ W[j-16+6] ^ W[j-28+6] ^ W[j-32+6]), 2);
+         W[j+7] = rotate_left((W[j-6+7] ^ W[j-16+7] ^ W[j-28+7] ^ W[j-32+7]), 2);
          }
 #endif
~~~~

With this, I get the 2.7 % slower execution.

I have not measured it but I think the original change might be beneficial compared the plain not-unrolled loop.

**Comment by webmaster128** [comment]
Thanks @wilx for digging into this and sharing your stuff. Further contributions are highly welcome.
